Eliminate the need for explicit RPMTEST_INIT

pmatilai opened the pull request:
See commits for details, but the main deal here is to reduce the boilerplate needed for tests and nudge test-authors towards not using (expensive) mutable snapshots unnecessarily.

dmnks commented on tests/README.md:
```diff
@@ -135,10 +135,10 @@ For the typical structure of a single test, consult GNU Autotest's
-* Use `RPMTEST_INIT` to create a mutable snapshot (optional)
+* Use `RPMTEST_SETUP` instead of `AT_SETUP` to run a test in an immutable
+  system image with writable `.` and `/tmp`.
+* Use `RPMTEST_SETUP_RW` instead of `AT_SETUP` to run a test in a mutable
```
These could be worded better: the peculiarity here is that test script itself still runs *in* the immutable environment even if using `RPMTEST_SETUP_RW`. As it is now, the text makes you think the script runs in the snapshot (i.e. that `touch /hello` would work) :sweat_smile: 

How about something like:

* Use `RPMTEST_SETUP` instead of `AT_SETUP`. If your test needs to alter the root filesystem (e.g. to install a package), use `RPMTEST_SETUP_RW` instead to request a mutable snapshot of the filesystem.

Or something of that sorts, you get the point (it's getting late here, my creative mind is tired :sleeping:).

pmatilai replied:
Oh right, I forgot about this detail that you still need runroot to actually write to it. 

Maybe it's enough to say "RPMTEST_SETUP_RW instead to prepare  the use of a mutable snapshot", followed by the bullet list on how to actually run stuff there. 

...

Except that ... runroot() and runroot_other() are getting incredibly pointless in this picture. We should hide that part into RPMTEST_CHECK() really, similarly to how we're fading the other bits into RPMTEST_SETUP* here. But that's a PR for another time :see_no_evil:

dmnks replied:
> Maybe it's enough to say "RPMTEST_SETUP_RW instead to prepare the use of a mutable snapshot", followed by the bullet list on how to actually run stuff there.

Oh yup, "prepare" sounds good to me, too, it's even better.

> Except that ... runroot() and runroot_other() are getting incredibly pointless in this picture. We should hide that part into RPMTEST_CHECK() really, similarly to how we're fading the other bits into RPMTEST_SETUP* here. But that's a PR for another time 🙈

Yep, I agree that's kinda the end goal here. We want this to be fully transparent to the user writing a test. The user should just call "rpm" etc. normally, without remembering to use "runroot" etc. (also see #2773).

dmnks replied:
That said, this can't be fully transparent, by design. The fact that the test logic runs *outside* the root filesystem (i.e. snapshot) subjected to testing is a feature, in fact. It ensures that the test logic itself isn't affected by itself. And at that point, the test author needs to realize *what* actually runs inside the snapshot and what doesn't.

dmnks replied:
... so I guess it's not entirely clear we want to get rid of the runroot prefix after all, this needs more thinking I guess :smile: But yup, something to ponder about as we go.

pmatilai replied:
The test author needs to be aware of where stuff runs of course, but the inside/outside difference is largely a historical remnant. For most test, there's no need for them to be aware of the "outside" anymore, and also most tests using runroot do so for all of their commands. So it might be just as well wrapped on the RPMTEST_CHECK() level (eg with RPMTEST_CHECK_RW())

dmnks replied:
What you mean is basically run the whole of RPMTEST_CHECK() *inside* the snapshot (if enabled with `RPMTEST_SETUP_RW`). Yes, that would simplify things conceptually, but would remove the ability to inspect and/or verify the results of a test from outside, like said, plus might turn out to be challening since a lot of the current rpmtests logic relies on this separation.

Anything in between is just obfuscating the fact that there *is* a split, I'm afraid :sweat_smile:

dmnks replied:
That said, this separation is only useful in theory, I think, since a test typically runs "rpm" (in the snapshot) to verify the results, anyway.

It might prove to be useful when there's some kind of destructive thing going on in the test, in which case you want to be able to "post-mortem" analyze the filesystem, but 99% of the tests may just get away without it.

pmatilai replied:
Yeah, running the whole of RPMTEST_CHECK inside the snapshot is what I mean. There may well be tests that fundamentally need access from the outside, and I don't think we should *remove* that capability, just that the vast majority of tests could be further simplified a lot by just running the whole thing inside.

dmnks replied:
So... thinking about it more, yep, no point in optimizing for this theoretical one or two tests when the majority of them are fine without the separation. In that case, just running the whole test script in the snapshot container might just be the best thing. But this is indeed for a separate PR/discussion so I'll shut up now :smile:

pmatilai replied:
Should probably lift this subtopic to a ticket though :laughing:

dmnks replied:
I can create one, have some more things to note on that topic as well :smile:

dmnks commented on tests/README.md:
```diff
@@ -156,6 +159,8 @@ as well as the existing tests.  Below are the specifics of RPM's test-suite:
       to the macro, e.g. `RPMTEST_USER([user1, user2])`.  Then, use
       `runroot_user -n <name>` to run a binary as a specific user
 * Use `RPMDB_RESET` to reinitialize a snapshot to an empty rpmdb (avoid this)
+* Use `RPMTEST_SNAPSHOT_MOUNT` create a mutable snapshot inside a immutable
```
*an* immutable :smile: (love these nitpicks, so easy to, well, nitpick)

dmnks reviewed (commented):
Some documentation tweaks may be needed, otherwise looks good, thanks! This makes rpm test writing finally something that one can actually grasp without deep prior knowledge :smile:

dmnks commented:
GH still doesn't support commenting on individual commit messages but the second commit in this series has a funny typo (emphasis mine):

> We'll need this in the following commits, **but** functional change here.

pmatilai commented:
Fixed the spotted typos/thinkos + adjusted the docs a bit. Also added a mention about building packages needing _RW for all the wrong reasons :sweat_smile:
